scripts/stav-covid.py

```diff
@@ -0,0 +1,153 @@
+
+# NOTE(rye): There is no filtering to prevent loading tables that aren't valid,
+# so please double-check the output of this script.
+tables = output.find_all("table")
+
+# A regex to match strings like "Sunday, August 23"
+date_re = re.compile('\w+, \w+ \d{1,2}')
+
+# A regex used to match timespec delimiters (e.g. the "-"" in "9-10:45")
+timespec_re = re.compile('\s*-\s*|\s*–\s*')
+
+# If the given tag only has one set of contents, just grab the contents.
+def meal_name_from(th):
+	if len(th.contents) == 1:
+		return th.get_text()
+	else:
+		return None
+
+# Attempt to parse a lazy timestring (either hh or hh:mm) to a datetime.  Note
+# that this datetime is context-free, and needs to be adjusted to make sense.
+def timeparse(string, date=None):
+	parts = string.split(":")
+	if len(parts) == 1:
+		return datetime.datetime.combine(date, datetime.time(hour=int(parts[0])))
+	elif len(parts) == 2:
+		return datetime.datetime.combine(date, datetime.time(hour=int(parts[0]), minute=int(parts[1])))
+	else:
+		return None
+
+# Start with an empty schedule set
+schedules = {}
+
+# For each of the HTML tables we found,
+for table in tables:
+	print("Processing a table...")
+
+	# Find the prior sibling, which is an h3 containing the day in some format.
+	# Each table should have an h3 before it that has date information in it.
+	header = table.find_previous_sibling("h3").get_text().strip()
+	print('Found a header sibling with text "{}"'.format(header))
+
+	# Search the text for just the date
+	matching_text = date_re.match(header).group(0)
+
+	# NOTE(rye): Parses text like "Sunday, August 23"
+	date = datetime.datetime.strptime(matching_text, "%A, %B %d")
+
+	# Move the date to the current year
+	date = datetime.datetime(date.today().year, date.month, date.day)
+
+	# If the computed date is more than six months before today, assume it actually belongs in the next year.
+	if date < date.today() + datetime.timedelta(days=-180):
+		date = datetime.datetime(date.year + 1, date.month, date.day)
+
+	# We only know the date, so let's back out to that.
+	date = date.date()
+
+	print("Inferred date of {}; processing table now.".format(date))
+
+	body = table.find("tbody")
+
+	# NOTE(rye): Should only have one thead
+	meals = [meal_name_from(th) for th in table.find("thead").find("tr").find_all("th")]
+
+	for idx, meal in enumerate(meals):
+		if idx == 0:
+			continue
+		for row in body.find_all("tr"):
+			cols = [col.get_text() for col in row.find_all("td")]
+			if len(cols) != len(meals):
+				continue
+			if len("".join(cols)) == 0:
+				continue
+			dorm = cols[0]
+			# meal = meal
+			timespec = cols[idx]
+
+			print("Processing table entry: {} gets {} at {} on {}".format(dorm, meal, timespec, date))
+
+			# Split the timespec on the timespec_re matches
+			times = [time.strip() for time in timespec_re.split(timespec)]
+
+			if len(times) != 2:
+				warnings.warn("timespec {} did not split nicely".format(timespec))
+				continue
```

Comment:
Cool, very happy to see this handling here considering there is a case in the data that breaks the regex

> UserWarning: timespec 10:15-10-45 did not split nicely

What are your thoughts on compensating? Does it make our program too brittle if we try to catch this case?

I'm thinking about how `timespec_re` matches for delimiters, and how timeparse splits on colons. If we find a dash where a colon _should_ be (that sounds bad as I type this), we could make a swap?

Eh. My idea sounds super brittle. Perhaps the happy path leaves less room for errors and thorny edge cases of further typos breaking the flow.

Reply:
I was thinking it might be smart to limit the number of splits via `maxsplit` on [`re.split`](https://docs.python.org/3/library/re.html#re.split), (then further splitting any invalid things on word boundaries and rejoining with a `:` before finally giving up) to catch stuff like that.

If you feel like I should, I totally could!

Reply:
That's a decent algorithm. I'm okay with it as-is, but I wonder what @hawkrives thinks?

Reply:
I'm good with either!

Reply:
I think I'll probably just opt to expect the happy path. If anyone complains, we do have the valid excuse that what was entered into the table is _technically_ wrong, and hopefully on the next pass it should get fixed. 😉